### ISSUE TITLE
Problem with backtracking step length calculations

### ISSUE THREAD

**Issue by aringh** [odlgroup/odl#344]
I think the backtracking step length calculations could need some caution.

1. If f(x) is finite, but f(x + a * direction) = infinite, and there is not enough iterations (before one hits the maximum number) in order to make f(x + a * direction) finite, this will cause problems for the user.

2. If f(x) = NaN, caution is also important. I'm not sure how NaN is interpreted, but for example
```
>>> np.nan >= 1
False
>>> np.nan <= 1
False
```
so if f(x + a*direction) = NaN, it is never going to enter the loop to make the step length smaller.

Edit: I have a fix to this, just need to find it. Then I will push it.

**Comment by adler-j** [comment]
I wrote this in the pull request, but my opinion is that we should be vary to enforce fixes here. In general, feeding a NaN to a function expecting a float is regarded as an exception. Perhaps for some applications it is reasonable to interpret NaN as Inf (thus shortening the step until it is OK), but in other it is not. I.m.o, a function returning NaN should be interpreted as an error, and we should throw some reasonable exception. It is then up to the user to fix this inside the functional, i.e. add this before returning `result`

```python
result = float('Inf') if result == float('Nan') else result
```

**Comment by aringh** [comment]
Your pull request in #348 looks fine to me. It gives clear answers to problems encountered.

**Comment by aringh** [comment]
I think this was closed via #348
